Refactor stock exit for performance and clarity

jniles opened the pull request:
The following major improvements are coming in this PR:

1. A rewritten `bhStockExitType` component that is smaller, cleaner, and has fewer bindings.  The component is now responsible for rendering its own label, rather than having multiple paths through the stock exit controller to render the label.  In the long run, we should have separate `bhStockExitType` and `bhStockEntryType` components.
2. The form logic is moved into an AngularJS service `StockExitFormService`.  This allows us to test things like the validation logic.
3. A new `LotItemService` is implemented for each grid row in the StockExit grid.  This allows us to handle the per-lot validation and the form validation a bit more cleanly.
4. Only a single call is made to the server for the stock lots!  Further, this route is rewritten to only JOIN on tables that we need and do the JOINs in a more optimal fashion.
5. I've written tests for the validation logic in the `LotItemService`.

### TODO
- [x] Finish the logic for removing items + revalidating the grid.
- [x] Finish and document the validation logic for the grid.  
    - [x] Add "actions" log under description showing the outstanding actions needed to be taken by a user to complete validation.
- [ ] Figure out how requisitions are fulfilled - right now it looks like the client is responsible for validating if a requisition has been fulfilled and that isn't probably correct.  Perhaps rewrite this portion of the code and move it server-side.  (See https://github.com/IMA-WorldHealth/bhima/issues/6424)
- [x] Implement `getLotsFromInventory()` feature to bring in lots from:
   - [x] Patient Invoice
   - [x] Service (requisition)
   - [x] Depot (requisition)
- [x] Submission
    - [x] Submit from Patient
    - [x] Submit from Loss
    - [x] Submit from Depot/Transfer
    - [x] Submit from Service
- [x] Rewrite description rendering logic - potentially in a `StockExitFormHelperService` to reduce file sizes.
- [x] Re-implement barcode scanning feature.
    - [x] Bring up the barcode scanning modal for the items
    - [x] Only choose from items in stock in the current depot.
    - [x] Display a prominent warning if the items are in stock in another depot. 
- [x] Write tests for LotItemService validation
- [ ] Write tests for StockExitFormService
- [X] Re-implement export to excel from grid

Closes #5577.

jniles committed:
feat(bhStockExitType): add simpler component

This commit implements a simpler stock exit type component.  It takes
care of the view without having the `reference` information passed back
from the StockExitController.  This allows us to reduce the code in the
StockExitController.

Progress towards #5577.

jniles committed:
chore(db): make sql more readable

This commit makes the SQL more readable.

jniles committed:
feat(LotItemService): implements lot item service

This commit implements the LotItemService to house lots on the client
and perform basic validation against them.  Unit tests are implemented
to ensure the behavior is correct.

jniles committed:
chore(StockExit): improve form validation

This commit encorporates common error cases into the LotItemService to
be used by the StockExitForm.

jniles committed:
feat: perfect error handling

This commit perfects the error handling of the LotItemService and the
LotForm service.

jniles committed:
fix(bhAddItem): remove unused barcode prop

Somehow a barcode binding was added to the bhAddItem.  This commit
removes it.

jniles committed:
chore(depots): add tracking expiration filter

When searching for an expiration date, we now include the
tracking_expiration flag to ensure only stock with tracked expiration
dates are matched.

jniles committed:
wip(StockExitForm): initial commit of SEFService

This is the initial commit of the StockExitFormService to document where
this is going.  The work on the bhExitType is completed and most of the
logic is moved into the StockExitFormService.  The clear logic has also
been implemented.

This is a work-in-progress.

Addresses #5577.

jniles committed:
test(StockExitFormHelper): add unit tests

Adds unit tests for the description rendering of the
StockExitFormHelper.  Note the testing $translate is a little finicky,
but this could be improved later.

jniles committed:
perf(depots): reduce expired/stock out calls

This commit reduces the number of stock out and expired stock calls made
by the components checking for stock out by caching the results on the
client.

jniles committed:
wip(stock): add error validation to stock exit

This commit adds per-line validation to the stock exit grid, as well as
nice tooltip rendering of the errors on hover.  All errors are properly
detected and displayed.

jniles committed:
wip(stock): exit form validation works

This commit fixes the exit form line-by-line validation so it works
as expected.  Each line self-validates with each condition when a change
is made.  We are also able to add/remove lines as needed.

jniles committed:
wip(stock): added info messages to exit

Adds informational messages to the stock exit form and an additional
validation criteria for allowing stock exits when expired.

jniles committed:
wip(stock): add pending actions log

This commit adds the pending actions log with details on the grid
validation.  It also tears out the stock_value dependency on the data
loaded by the grid - this can be computed in real time as data is being
inserted into the database.

jniles committed:
wip(stock): exit to loss and depot work

This commit finalizes the rendering for stock exit to loss and depot.
The translation keys still need some work.

jniles committed:
wip(stock): fix validation on removeItem()

Ensures validation happens when we remove the items form the grid.

jniles committed:
fix(stock): bhStockOut indicator dates

Fixes the date parsing in the bhStockOut component.

jniles committed:
wip(stock): submissions all work

Submission from each different route works as expected.

jniles committed:
wip(stock): grid export works again

jmcameron commented on client/src/js/components/bhStockExpired/bhStockExpired.js:
```diff
@@ -9,33 +9,40 @@ angular.module('bhima.components')
+
+    $q.all([
+      fetchExpiredStock(),
+      getDepot(),
```
👍 Good to put this into the promises

jmcameron commented on client/src/js/components/bhStockOut/bhStockOut.js:
```diff
@@ -49,8 +48,8 @@ function bhStockOutController(Depots, moment, Notify, $filter, $q) {
         inventories.forEach(inventory => {
-          inventory.stock_out_date_raw = $date(inventory.stock_out_date);
-          inventory.stock_out_date_parsed = moment(inventory.stock_out_date).fromNow();
+          inventory.stock_out_date_raw = $date(inventory.date);
```
Why change inventory.stock_out_date to inventory.date?

jniles replied:
To address a bug - the server sends back the `date` now, not the `stock_out_date`.  This is a bug in master, and I'm fixing it here, since I had to investigate it to see if it was me that broke it.

jmcameron commented on client/src/modules/stock/LotItem.service.js:
```diff
@@ -0,0 +1,366 @@
+      delete clone.tracking_consumption;
+    }
+
+    if (clone.tracking_expiration !== undefined) {
```
This pattern seems repetitive.  Is there any way to streamline it?

jniles replied:
I can't think of a shorter version... can you?

jmcameron commented on client/src/modules/stock/LotItem.service.js:
```diff
@@ -0,0 +1,366 @@
+   * @function hasLotInformation
+   *
+   * @description
+   * Checks if the lot information is availability.
```
Typo? (last word on line)

jmcameron commented on client/src/modules/stock/LotItem.service.js:
```diff
@@ -0,0 +1,366 @@
+   *    override the logic here without overriding the error handling.
+   * 3) If a date comparison is passed in, the lot's expiration state is computed by
+   *    comparing to the comparison date.  If the lot's expiration_date  is before the
+   *    comparison date, it is considered expired.  If the lot's expiration_date is after,
```
what about "the same as" case?

jmcameron commented on client/src/modules/stock/LotItem.service.js:
```diff
@@ -0,0 +1,366 @@
+    // (for example, if this is a past date).  If not, it will validate
+    // against the current date.
+    } else {
+      isLotExpired = this.expiration_date < comparisonDate;
```
👍

jmcameron commented on client/src/modules/stock/LotItem.service.js:
```diff
@@ -0,0 +1,366 @@
+   * is empty by default (no quantity in stock), no error is triggered.
+   */
+  Lot.prototype.isUnused = function isUnused() {
+    const hasEmptyQuantityAndUnused = this.isEmpty() && this._quantity_available !== 0;
```
Can we ever have a case where _quantity_available is negative?   If so, then we should use: > 0?

jniles replied:
Yes, we can, but by accident.  In several databases we have observed where the quantity in stock is negative, but it isn't supposed to happen.

jmcameron replied:
I realize that this is not supposed to happen.  But shouldn't we prevent that type of error from affecting this functionality by using '> 0' instead of '!== 0'?

jmcameron commented on client/src/modules/stock/LotItem.service.js:
```diff
@@ -0,0 +1,366 @@
+   * Returns true if the lot is consumable.
+   */
+  Lot.prototype.isConsumable = function isConsumable() {
+    return this.consumable === 1;
```
Or  use the !! syntax?

jmcameron commented on client/src/modules/stock/StockModal.service.js:
```diff
@@ -40,7 +41,6 @@ function StockModalService(Modal) {
     const params = angular.extend(modalParameters, {
       templateUrl  : 'modules/stock/entry/modals/generateAssetTags.modal.html',
       controller   : 'GenerateTagsModalController',
-      controllerAs : '$ctrl',
```
Why were these removed?  Is there a default?

jniles replied:
Yes, I added it to the default `modalParameters`.

jmcameron reviewed (commented):
I reviewed the code and left inline comments.

I tested it.  For the most part, the new Stock Exit works very nicely and is friendlier to the user.  I like the "Pending actions" section.

A few issues:
- If I add an inventory item, it selects the first available lot and inserts its quantity correctly.  However, if you at that point decide you want a different lot, when you select it, the quantity is zero (it is not updated).
- I tried doing an exit to Depot and selecting the inventory item worked fine, but the client side had errors when I tried to submit: TypeError: Cannot read properties of undefined (reading 'data')
- If I have started with "Loss", added an item, and decide I should do exit to "Depot", the inventory item grid should be erased (currently it is not).
- When I tried an exit to Service, and added an inventory item, it picked a lot that was expired.  I saw similar behavior for exit to Patent.

jniles committed:
wip(stock exit): clear grid on setExitType()

This commit make the grid get reset when a new exit type is selected.

jniles committed:
wip(stock exit): implement import from requisition

This commit implements the "import from requisition" feature for both
services and depots.

jniles committed:
chore(stock exit): lint fixes

jniles committed:
test(stock): fix stock tests

Fixes the stock tests for getting the stock in a depot.

jniles committed:
fix(stock exit): implement barcode scanning feat

Adds the barcode scanning feature back in so that users can scan the
barcode of a lot and add it to the grid.

mbayopanda reviewed (changes requested):
This is big refactoring and the code is very clear now, I just noticed that with the test database when we test exit on the "Quinine Bichlorhydrate, sirop, 100mg base/5ml, 100ml, flacon, Unité", we have these issues:

- The available quantity is set to zero (0) when the lot is changed
- We can submit the stock exit with zero as quantity

jniles committed:
fix(stock exit): set quantity_available

This commit fixes an issue with setting the quantity available when the
lot is switched

jniles committed:
i18n(stock): add French translations for stock exit

This commit adds French translations for stock exit.

jniles committed:
perf(stock): improve createMovement speed

This improves the performance of createMovement speed by using the
updated "getLotsForDate" function.

jniles committed:
perf(stock): fix the exit helper

Improves performance during stock movement creation and lookup in the
helpers.

jmcameron commented on client/src/js/services/util.js:
```diff
@@ -27,6 +27,11 @@ function UtilService(moment) {
     return null;
   };
 
+  service.getUniqueBy = function getUniqueBy(array, prop) {
```
Please add a descriptive comment for this function

jmcameron reviewed (commented):
I took a brief glance at the code again.

I tested Stock Exit with bhima_test
- In Exit to stock to patient
   - once I selected a patient and an invoice, it (correctly) complained that
     one item was out of stock and would be skipped.  Excellent message!
   - If the invoice calls for an expired stock, it is still listed and the
     warning message is not very clear.  A warning (in the Pending Actions
     section) could say: This stock is expired, select another lot.  (or
     something similar).
   - Good message when an invoice could only partially be filled.
- When I tried Exit to Service, I selected a stock, etc, but clicking on
  the [Submit] button did nothing.
- Exit to Depot worked fine
- Exit to Loss worked fine

jniles committed:
fix(stock exit): prevent duplicates

This commit fixes the duplicate selection of lots.

jniles committed:
fix(stock): exit to service typo

Fixed a typo in the "exit to service" description rendering.

Closes #5577.

mbayopanda reviewed (approved):
Good job

LGTM 👍🏽

jniles committed:
chore: bump deps
